synapseclient/models/activity.py

```diff
@@ -1 +1,345 @@
+        name: The name of the URL.
+        url: The external URL of the file that was used such as a link to a
+            GitHub commit or a link to a specific version of a software tool.
+    """
+
+    name: Optional[str] = None
+    """The name of the URL."""
+
+    url: Optional[str] = None
+    """The external URL of the file that was used such as a link to a GitHub commit
+    or a link to a specific version of a software tool."""
+
+
+# TODO: When Views and Datasets are added we should add Activity to them.
+@dataclass
+class Activity:
+    """
+    An activity is a Synapse object that helps to keep track of what objects were used
+    in an analysis step, as well as what objects were generated. Thus, all relationships
+    between Synapse objects and an activity are governed by dependencies. That is, an
+    activity needs to know what it `used`, and outputs need to know what activity
+    they were `generatedBy`.
+
+    Attributes:
+        id: The unique immutable ID for this actvity.
+        name: A name for this Activity.
+        description: A description for this Activity.
+        etag: Synapse employs an Optimistic Concurrency Control (OCC) scheme to handle
+            concurrent updates. Since the E-Tag changes every time an entity is updated it
+            is used to detect when a client's current representation of an entity is out-of-date.
+        created_on: The date this object was created.
+        modified_on: The date this object was last modified.
+        created_by: The user that created this object.
+        modified_by: The user that last modified this object.
+        used: The entities or URLs used by this Activity.
+        executed: The entities or URLs executed by this Activity.
+    """
+
+    id: Optional[str] = None
+    """The unique immutable ID for this actvity."""
+
+    name: Optional[str] = None
+    """A name for this Activity."""
+
+    description: Optional[str] = None
+    """A description for this Activity."""
+
+    etag: Optional[str] = None
+    """Synapse employs an Optimistic Concurrency Control (OCC) scheme to handle
+    concurrent updates. Since the E-Tag changes every time an entity is updated it
+    is used to detect when a client's current representation of an entity is out-of-date."""
+
+    created_on: Optional[str] = None
+    """The date this object was created."""
+
+    modified_on: Optional[str] = None
+    """The date this object was last modified."""
+
+    created_by: Optional[str] = None
+    """The user that created this object."""
+
+    modified_by: Optional[str] = None
+    """The user that last modified this object."""
+
+    used: List[Union[UsedEntity, UsedURL]] = field(default_factory=list)
+    """The entities used by this Activity."""
+
+    executed: List[Union[UsedEntity, UsedURL]] = field(default_factory=list)
+    """The entities executed by this Activity."""
+
+    def fill_from_dict(self, synapse_activity: Synapse_Activity) -> "Activity":
```

Comment:
I'm a little confused with the `synapse_activity` type and the description.  If its response from API, would it be `dict`?  I see that you're also supporting the return object of `syn.getActivity`

Reply:
In the `client.py` the save action calls down to this method:
```
    def _saveActivity(self, activity: Activity) -> Activity:
        """
        Save the Activity

        Arguments:
            activity: The Activity to be saved

        Returns:
            An Activity object
        """
        if "id" in activity:
            # We're updating provenance
            uri = "/activity/%s" % activity["id"]
            activity = Activity(data=self.restPUT(uri, json.dumps(activity)))
        else:
            activity = self.restPOST("/activity", body=json.dumps(activity))
        return activity
```


Specifically the return type is `Activity`: https://github.com/Sage-Bionetworks/synapsePythonClient/blob/SYNPY-1344-activity-model/synapseclient/activity.py#L133

I want to get away from having this 

1. `from synapseclient.models.activity import Activity` and
2. `from synapseclient.activity import Activity as Synapse_Activity`

So we don't have 2 models. In the end though I need an exact object of what the REST API is expecting/returning, and what the OOP model looks like; A translation between the two. I am hoping do something with this in https://sagebionetworks.jira.com/browse/SYNPY-1411 - Or perhaps when the open API generated client drops I am use that and automatically get "what the REST API is expecting/returning", and write keep the translation between the 2 in the current python client (This translation is what the `fill_from_dict` and [this line is doing](https://github.com/Sage-Bionetworks/synapsePythonClient/blob/SYNPY-1344-activity-model/synapseclient/models/activity.py#L239))

Reply:
Got it, I think it was confusing from the function name since `fill_from_dict` seems like it will take in a `dict`.  Although... the function does look like it would support a `dict. ha.

Reply:
I suppose I could add a union type hint here too